File: release_it/chapter5/section3_code_examples.py
#!/usr/bin/env python3
"""
Section 3: Python Code Examples - Infrastructure Variability Handling

This module demonstrates patterns for handling infrastructure variability
as discussed in Chapter 5 of "Release It!"

These examples complement the Go examples - Python is often used for
observability, monitoring, and automation scripts.
"""

import asyncio
import contextlib
import random
import time
from dataclasses import dataclass, field
from typing import Callable, Optional
from functools import wraps
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


# =============================================================================
# EXAMPLE 1: Infrastructure Metrics Collection
# =============================================================================

@dataclass
class InfrastructureMetrics:
    """
    Staff-level insight: Infrastructure metrics are often the first indicator
    of problems that will later manifest as application errors.
    """
    cpu_steal: float = 0.0
    io_wait: float = 0.0
    memory_usage: float = 0.0
    network_latency: float = 0.0
    disk_latency: float = 0.0

    # Historical data for trend analysis
    history: list = field(default_factory=list)

    def record(self):
        """Record current metrics to history"""
        self.history.append({
            'timestamp': time.time(),
            'cpu_steal': self.cpu_steal,
            'io_wait': self.io_wait,
            'memory_usage': self.memory_usage,
            'network_latency': self.network_latency,
            'disk_latency': self.disk_latency,
        })

        # Keep only last 1000 samples
        if len(self.history) > 1000:
            self.history.pop(0)

    def get_alert_status(self) -> list:
        """
        Determine if any metrics exceed thresholds.
        Staff-level insight: These thresholds should be tuned to your workload.
        """
        alerts = []

        if self.cpu_steal > 10:
            alerts.append(f"CRITICAL: CPU steal at {self.cpu_steal:.1f}% (threshold: 10%)")

        if self.io_wait > 20:
            alerts.append(f"WARNING: I/O wait at {self.io_wait:.1f}% (threshold: 20%)")

        if self.memory_usage > 85:
            alerts.append(f"WARNING: Memory usage at {self.memory_usage:.1f}% (threshold: 85%)")

        if self.network_latency > 100:
            alerts.append(f"WARNING: Network latency at {self.network_latency:.1f}ms (threshold: 100ms)")

        if self.disk_latency > 50:
            alerts.append(f"WARNING: Disk latency at {self.disk_latency:.1f}ms (threshold: 50ms)")

        return alerts

    def is_healthy(self) -> bool:
        """Check if infrastructure is operating normally"""
        return (
            self.cpu_stele < 5 and
            self.io_wait < 10 and
            self.memory_usage < 80 and
            self.network_latency < 50 and
            self.disk_latency < 20
        )


# =============================================================================
# EXAMPLE 2: Simulated Infrastructure Variability
# =============================================================================

class SimulatedInfrastructure:
    """
    Simulates the variable performance of virtualized infrastructure.
    Staff-level insight: Use this to test your application's resilience patterns.
    """

    def __init__(
        self,
        base_latency_ms: float = 10,
        variance_ms: float = 50,
        failure_rate: float = 0.05,
        noisy_neighbor_probability: float = 0.1
    ):
        self.base_latency_ms = base_latency_ms
        self.variance_ms = variance_ms
        self.failure_rate = failure_rate
        self.noisy_neighbor_probability = noisy_neighbor_probability
        self.is_noisy = False

    def simulate_call(self, operation_name: str) -> float:
        """
        Simulate a call with variable latency.
        Returns the actual latency in milliseconds.
        """
        # Simulate noisy neighbor effect (random "bad days")
        if random.random() < self.noisy_neighbor_probability:
            self.is_noisy = True
            # Noisy neighbor causes massive latency spikes
            latency = self.base_latency_ms + random.uniform(0, self.variance_ms * 5)
        else:
            self.is_noisy = False
            # Normal variability (context switching, etc.)
            latency = self.base_latency_ms + random.uniform(0, self.variance_ms)

        # Simulate failures (hardware issues, timeouts)
        if random.random() < self.failure_rate:
            raise ConnectionError(f"Infrastructure failure during {operation_name}")

        # Simulate VM migration (occasional huge spike)
        if random.random() < 0.01:
            latency += random.uniform(500, 2000)  # 0.5-2 second spike

        return latency


# =============================================================================
# EXAMPLE 3: Timeout with Asyncio
# =============================================================================

async def async_operation_with_timeout(coro, timeout_seconds: float):
    """
    Staff-level insight: In async Python, you must explicitly handle timeouts.
    This is even more critical when infrastructure is unreliable.
    """
    try:
        return await asyncio.wait_for(coro, timeout=timeout_seconds)
    except asyncio.TimeoutError:
        raise TimeoutError(f"Operation exceeded {timeout_seconds}s timeout")


async def unreliable_service_call(infrastructure: SimulatedInfrastructure):
    """
    Simulates calling an unreliable service.
    Staff-level insight: This is what happens when you call a service
    running on a poorly-provisioned VM.
    """
    latency = infrastructure.simulate_call("service_call")
    await asyncio.sleep(latency / 1000)  # Convert ms to seconds
    return f"completed in {latency:.1f}ms"


# =============================================================================
# EXAMPLE 4: Retry with Exponential Backoff
# =============================================================================

class RetryConfig:
    """Configuration for retry behavior"""

    def __init__(
        self,
        max_retries: int = 3,
        base_delay: float = 0.1,  # 100ms
        max_delay: float = 5.0,   # 5 seconds
        exponential_base: float = 2.0,
        jitter: bool = True
    ):
        self.max_retries = max_retries
        self.base_delay = base_delay
        self.max_delay = max_delay
        self.exponential_base = exponential_base
        self.jitter = jitter


def with_retry(config: RetryConfig):
    """
    Decorator that adds retry with exponential backoff to async functions.
    Staff-level insight: Jitter prevents thundering herd when many clients retry.
    """
    def decorator(func: Callable):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            last_exception = None

            for attempt in range(config.max_retries + 1):
                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    last_exception = e

                    if attempt < config.max_retries:
                        # Calculate delay with exponential backoff
                        delay = min(
                            config.base_delay * (config.exponential_base ** attempt),
                            config.max_delay
                        )

                        # Add jitter to prevent thundering herd
                        if config.jitter:
                            delay = delay * (0.5 + random.random())

                        logger.warning(
                            f"Attempt {attempt + 1} failed: {e}. "
                            f"Retrying in {delay:.2f}s..."
                        )
                        await asyncio.sleep(delay)
                    else:
                        logger.error(f"All {config.max_retries + 1} attempts failed")

            raise last_exception

        return wrapper
    return decorator


# =============================================================================
# EXAMPLE 5: Circuit Breaker
# =============================================================================

class CircuitBreaker:
    """
    Circuit breaker pattern for handling infrastructure failures.
    Staff-level insight: When infrastructure is failing, fail fast to preserve
    resources and allow recovery.
    """

    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"

    def __init__(
        self,
        failure_threshold: int = 5,
        success_threshold: int = 2,
        timeout: float = 30.0
    ):
        self.failure_threshold = failure_threshold
        self.success_threshold = success_threshold
        self.timeout = timeout
        self.state = self.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time: Optional[float] = None

    async def call(self, func: Callable, *args, **kwargs):
        """Execute function through circuit breaker"""

        # Check if we should transition from OPEN to HALF_OPEN
        if self.state == self.OPEN:
            if time.time() - self.last_failure_time > self.timeout:
                logger.info("Circuit breaker transitioning to HALF_OPEN")
                self.state = self.HALF_OPEN
                self.success_count = 0
            else:
                raise ConnectionError("Circuit breaker is OPEN")

        try:
            result = await func(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure()
            raise e

    def _on_success(self):
        """Handle successful call"""
        if self.state == self.HALF_OPEN:
            self.success_count += 1
            if self.success_count >= self.success_threshold:
                logger.info("Circuit breaker closing after successful recovery")
                self.state = self.CLOSED
                self.failure_count = 0
        elif self.state == self.CLOSED:
            self.failure_count = 0  # Reset on success

    def _on_failure(self):
        """Handle failed call"""
        self.failure_count += 1
        self.last_failure_time = time.time()

        if self.state == self.HALF_OPEN:
            logger.warning("Circuit breaker reopening after failed half-open attempt")
            self.state = self.OPEN
        elif self.failure_count >= self.failure_threshold:
            logger.warning(f"Circuit breaker opening after {self.failure_count} failures")
            self.state = self.OPEN


# =============================================================================
# EXAMPLE 6: Health Check with Infrastructure Metrics
# =============================================================================

class InfrastructureHealthChecker:
    """
    Performs health checks that include infrastructure metrics.
    Staff-level insight: Traditional health checks often don't catch
    infrastructure problems until they cause failures.
    """

    def __init__(self, infrastructure: SimulatedInfrastructure):
        self.infrastructure = infrastructure
        self.metrics_history = []

    async def perform_health_check(self) -> dict:
        """
        Perform a comprehensive health check including infrastructure metrics.
        Returns detailed status for monitoring and alerting.
        """
        check_result = {
            "healthy": True,
            "infrastructure_issues": [],
            "recommendations": [],
            "timestamp": time.time()
        }

        # Simulate collecting infrastructure metrics
        metrics = InfrastructureMetrics()
        metrics.cpu_steal = random.uniform(0, 15)  # Simulated
        metrics.io_wait = random.uniform(0, 25)
        metrics.memory_usage = random.uniform(30, 90)
        metrics.network_latency = random.uniform(5, 150)
        metrics.disk_latency = random.uniform(2, 60)

        # Check for issues
        alerts = metrics.get_alert_status()

        if alerts:
            check_result["healthy"] = False
            check_result["infrastructure_issues"] = alerts

            # Generate recommendations based on specific issues
            if metrics.cpu_steal > 10:
                check_result["recommendations"].append(
                    "Consider moving to dedicated hosts or reducing VM density"
                )

            if metrics.io_wait > 20:
                check_result["recommendations"].append(
                    "Review I/O intensive operations; consider SSD storage"
                )

            if metrics.network_latency > 100:
                check_result["recommendations"].append(
                    "Check for network contention; consider VPC peering"
                )

        # Store for trend analysis
        self.metrics_history.append(check_result)
        if len(self.metrics_history) > 100:
            self.metrics_history.pop(0)

        return check_result


# =============================================================================
# DEMO: Running the Examples
# =============================================================================

async def run_demo():
    """Demonstrate all patterns"""

    print("=" * 60)
    print("Chapter 5: Infrastructure Variability - Python Demo")
    print("=" * 60)
    print()

    # Create simulated infrastructure with high variability
    infra = SimulatedInfrastructure(
        base_latency_ms=10,
        variance_ms=100,  # High variance!
        failure_rate=0.2,  # 20% failure rate
        noisy_neighbor_probability=0.2  # 20% chance of "bad day"
    )

    # Demo 1: Metrics Collection
    print("--- Demo 1: Infrastructure Metrics ---")
    metrics = InfrastructureMetrics(
        cpu_steal=12.5,
        io_wait=22.3,
        memory_usage=87.0,
        network_latency=150.0,
        disk_latency=45.0
    )

    alerts = metrics.get_alert_status()
    for alert in alerts:
        print(f"  ⚠️  {alert}")
    print()

    # Demo 2: Retry with Backoff
    print("--- Demo 2: Retry with Exponential Backoff ---")

    @with_retry(RetryConfig(max_retries=3, base_delay=0.1))
    async def unreliable_call():
        latency = infra.simulate_call("retry_test")
        await asyncio.sleep(latency / 1000)
        return f"success after {latency:.1f}ms"

    try:
        result = await unreliable_call()
        print(f"  ✓ {result}")
    except Exception as e:
        print(f"  ✗ Failed: {e}")
    print()

    # Demo 3: Circuit Breaker
    print("--- Demo 3: Circuit Breaker ---")
    cb = CircuitBreaker(failure_threshold=3, timeout=2.0)

    # Simulate some failures
    for i in range(10):
        try:
            await cb.call(unreliable_call)
            print(f"  Request {i+1}: SUCCESS (state: {cb.state})")
        except ConnectionError as e:
            print(f"  Request {i+1}: BLOCKED - {e} (state: {cb.state})")
            if cb.state == CircuitBreaker.OPEN:
                print("  → Circuit is open, waiting for timeout...")
                break

    print()

    # Demo 4: Health Check
    print("--- Demo 4: Infrastructure Health Check ---")
    health_checker = InfrastructureHealthChecker(infra)

    for _ in range(3):
        result = await health_checker.perform_health_check()
        status = "✓ HEALTHY" if result["healthy"] else "✗ UNHEALTHY"
        print(f"  {status}")

        if result["infrastructure_issues"]:
            for issue in result["infrastructure_issues"]:
                print(f"    - {issue}")
        if result["recommendations"]:
            for rec in result["recommendations"]:
                print(f"    → {rec}")

        await asyncio.sleep(0.5)

    print()
    print("=" * 60)
    print("Demo complete!")
    print("=" * 60)


if __name__ == "__main__":
    asyncio.run(run_demo())

File: release_it/chapter5/section3_code_examples.go
package main

import (
	"context"
	"errors"
	"fmt"
	"math"
	"math/rand"
	"sync"
	"time"
)

// =============================================================================
// SECTION 3: ANNOTATED CODE EXAMPLES
// =============================================================================
// These examples demonstrate patterns for handling infrastructure variability
// as discussed in Chapter 5 of "Release It!"
// =============================================================================

// =============================================================================
// EXAMPLE 1: Handling Variable Latency with Timeout and Backoff
// =============================================================================

// ErrTimeout represents a timeout error
var ErrTimeout = errors.New("operation timed out")

// ErrCircuitOpen represents an open circuit breaker
var ErrCircuitOpen = errors.New("circuit breaker is open")

// InfrastructureClient simulates a client making requests to infrastructure
// that has variable latency due to virtualization/ hardware issues
type InfrastructureClient struct {
	name          string
	baseLatency   time.Duration
	variance      time.Duration // This represents the variability from virtualization
	failureRate   float64
	circuitOpen   bool
	failureCount  int
	successCount  int
	mu            sync.Mutex
}

// NewInfrastructureClient creates a client with specified characteristics
// Why: In production, you need to model infrastructure variability to test resilience
func NewInfrastructureClient(name string, baseLatency, variance time.Duration, failureRate float64) *InfrastructureClient {
	return &InfrastructureClient{
		name:        name,
		baseLatency: baseLatency,
		variance:    variance,
		failureRate: failureRate,
	}
}

// SimulateRequest makes a request with variable latency
// The variance simulates: CPU steal, I/O wait, network virtualization overhead
func (c *InfrastructureClient) SimulateRequest(ctx context.Context) (string, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.circuitOpen {
		return "", ErrCircuitOpen
	}

	// Simulate variable latency (the "un-virtualized ground" problem)
	// Staff-level insight: This is what happens when noisy neighbors consume resources
	actualLatency := c.baseLatency + time.Duration(rand.Float64()*float64(c.variance))

	// Check for timeout
	select {
	case <-ctx.Done():
		c.failureCount++
		return "", ctx.Err()
	default:
	}

	// Simulate the request
	time.Sleep(actualLatency)

	// Simulate random failures (hardware issues, network problems)
	// Staff-level insight: Random failures often come from hardware, not code
	if rand.Float64() < c.failureRate {
		c.failureCount++
		return "", errors.New("infrastructure failure")
	}

	c.successCount++
	return fmt.Sprintf("response from %s (latency: %v)", c.name, actualLatency), nil
}

// =============================================================================
// NAIVE APPROACH: No timeouts, no resilience
// =============================================================================

// NaiveRequestWithoutTimeout demonstrates what most developers do
// ❌ This will hang indefinitely when infrastructure has issues
func NaiveRequestWithoutTimeout(client *InfrastructureClient) (string, error) {
	// BAD: No context, no timeout
	// When the hypervisor throttles CPU or I/O waits spike, this hangs forever
	result, err := client.SimulateRequest(context.Background())
	if err != nil {
		return "", fmt.Errorf("request failed: %w", err)
	}
	return result, nil
}

// =============================================================================
// PRODUCTION APPROACH: Timeout with context
// =============================================================================

// RequestWithTimeout demonstrates proper timeout handling
// Why: Infrastructure variability means operations can take arbitrarily long
// Staff-level insight: Timeout is your circuit breaker at the request level
func RequestWithTimeout(client *InfrastructureClient, timeout time.Duration) (string, error) {
	// GOOD: Create context with timeout
	// This prevents hanging when VM migration or noisy neighbors cause delays
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	result, err := client.SimulateRequest(ctx)
	if err != nil {
		if errors.Is(err, context.DeadlineExceeded) {
			return "", fmt.Errorf("%w: infrastructure latency exceeded %v", ErrTimeout, timeout)
		}
		return "", fmt.Errorf("request failed: %w", err)
	}
	return result, nil
}

// =============================================================================
// EXAMPLE 2: Circuit Breaker Pattern
// =============================================================================

// CircuitBreakerState represents the state of a circuit breaker
type CircuitBreakerState int

const (
	CircuitClosed CircuitBreakerState = iota
	CircuitOpen
	CircuitHalfOpen
)

// CircuitBreaker prevents cascading failures from infrastructure issues
// Why: When infrastructure is failing, don't keep hammering it
// Staff-level insight: This is the application-level defense against bad hardware
type CircuitBreaker struct {
	mu               sync.Mutex
	state            CircuitBreakerState
	failureThreshold int
	successThreshold int
	failureCount     int
	successCount     int
	lastFailureTime  time.Time
	openDuration    time.Duration
}

// NewCircuitBreaker creates a circuit breaker with specified thresholds
func NewCircuitBreaker(failureThreshold int, openDuration time.Duration) *CircuitBreaker {
	return &CircuitBreaker{
		state:            CircuitClosed,
		failureThreshold: failureThreshold,
		openDuration:     openDuration,
	}
}

// Execute runs a function through the circuit breaker
// Staff-level insight: This pattern comes from electrical engineering - fail fast
// to prevent damage (in this case, cascading failures)
func (cb *CircuitBreaker) Execute(fn func() error) error {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case CircuitOpen:
		// Check if we should try half-open
		if time.Since(cb.lastFailureTime) > cb.openDuration {
			cb.state = CircuitHalfOpen
			cb.successCount = 0
		} else {
			return ErrCircuitOpen
		}
	case CircuitHalfOpen:
		// Allow limited requests to test recovery
	}

	err := fn()

	if err != nil {
		cb.recordFailure()
		return err
	}

	cb.recordSuccess()
	return nil
}

func (cb *CircuitBreaker) recordFailure() {
	cb.failureCount++
	cb.lastFailureTime = time.Now()

	if cb.state == CircuitHalfOpen {
		cb.state = CircuitOpen // Failed during test, go back to open
	} else if cb.failureCount >= cb.failureThreshold {
		cb.state = CircuitOpen // Too many failures, open the circuit
	}
}

func (cb *CircuitBreaker) recordSuccess() {
	cb.successCount++

	if cb.state == CircuitHalfOpen {
		if cb.successCount >= cb.successThreshold {
			cb.state = CircuitClosed
			cb.failureCount = 0
		}
	} else {
		// Reset failure count on success in closed state
		cb.failureCount = 0
	}
}

// GetState returns the current state of the circuit breaker
func (cb *CircuitBreaker) GetState() CircuitBreakerState {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	return cb.state
}

// =============================================================================
// EXAMPLE 3: Retry with Exponential Backoff
// =============================================================================

// RetryConfig holds configuration for retry behavior
type RetryConfig struct {
	MaxRetries     int
	InitialDelay   time.Duration
	MaxDelay       time.Duration
	Multiplier     float64
	ShouldRetry    func(error) bool
}

// DefaultRetryConfig returns sensible defaults for infrastructure retries
// Why: Infrastructure failures are often transient (temp CPU spike, brief I/O wait)
func DefaultRetryConfig() *RetryConfig {
	return &RetryConfig{
		MaxRetries:   3,
		InitialDelay: 100 * time.Millisecond,
		MaxDelay:     5 * time.Second,
		Multiplier:   2.0,
		ShouldRetry: func(err error) bool {
			// Don't retry on circuit open or explicit non-retryable errors
			if errors.Is(err, ErrCircuitOpen) {
				return false
			}
			return true
		},
	}
}

// RetryWithBackoff demonstrates retry with exponential backoff
// Staff-level insight: The backoff allows infrastructure to recover
// (VM migration completes, noisy neighbor's job ends, etc.)
func RetryWithBackoff(client *InfrastructureClient, cfg *RetryConfig) (string, error) {
	var lastErr error
	delay := cfg.InitialDelay

	for attempt := 0; attempt <= cfg.MaxRetries; attempt++ {
		if attempt > 0 {
			// Add jitter to prevent thundering herd
			jitter := time.Duration(rand.Float64() * float64(delay))
			fmt.Printf("Retry %d: waiting %v before retry...\n", attempt, delay+jitter)
			time.Sleep(delay + jitter)

			// Exponential backoff
			delay = time.Duration(math.Min(
				float64(delay)*cfg.Multiplier,
				float64(cfg.MaxDelay),
			))
		}

		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()

		result, err := client.SimulateRequest(ctx)
		if err == nil {
			return result, nil
		}

		lastErr = err

		// Check if we should retry this error
		if cfg.ShouldRetry != nil && !cfg.ShouldRetry(err) {
			return "", err
		}

		fmt.Printf("Attempt %d failed: %v\n", attempt+1, err)
	}

	return "", fmt.Errorf("all retries exhausted, last error: %w", lastErr)
}

// =============================================================================
// EXAMPLE 4: Graceful Degradation
// =============================================================================

// ServiceLevel represents different levels of service capability
type ServiceLevel int

const (
	LevelFull ServiceLevel = iota
	LevelDegraded
	LevelMinimal
	LevelDown
)

// DegradedService demonstrates graceful degradation
// Why: When infrastructure is struggling, reduce load rather than crash
// Staff-level insight: This is the "safety valve" pattern from the chapter
type DegradedService struct {
	client          *InfrastructureClient
	currentLevel    ServiceLevel
	requestsFull    int
	requestsDegraded int
	requestsMinimal int
	mu              sync.Mutex
}

// NewDegradedService creates a service with graceful degradation capability
func NewDegradedService(client *InfrastructureClient) *DegradedService {
	return &DegradedService{
		client:       client,
		currentLevel: LevelFull,
	}
}

// HandleRequest handles a request with appropriate degradation
// Staff-level insight: At LevelDegraded, we skip expensive operations
// At LevelMinimal, we return cached/stale data if possible
func (s *DegradedService) HandleRequest(ctx context.Context) (string, error) {
	s.mu.Lock()
	level := s.currentLevel
	s.mu.Unlock()

	switch level {
	case LevelFull:
		return s.handleFull(ctx)
	case LevelDegraded:
		return s.handleDegraded(ctx)
	case LevelMinimal:
		return s.handleMinimal(ctx)
	default:
		return "", errors.New("service unavailable")
	}
}

func (s *DegradedService) handleFull(ctx context.Context) (string, error) {
	s.mu.Lock()
	s.requestsFull++
	s.mu.Unlock()

	// Full processing - make the infrastructure call
	result, err := s.client.SimulateRequest(ctx)
	if err != nil {
		// If we fail repeatedly, degrade
		s.mu.Lock()
		if s.requestsFull > 5 {
			s.currentLevel = LevelDegraded
			fmt.Println("⚠️ Degrading to LevelDegraded")
		}
		s.mu.Unlock()
	}
	return result, err
}

func (s *DegradedService) handleDegraded(ctx context.Context) (string, error) {
	s.mu.Lock()
	s.requestsDegraded++
	s.mu.Unlock()

	// Skip expensive operations - maybe just cache check
	// For demo, we'll use a shorter timeout
	ctx, cancel := context.WithTimeout(ctx, 500*time.Millisecond)
	defer cancel()

	result, err := s.client.SimulateRequest(ctx)
	if err != nil {
		s.mu.Lock()
		s.currentLevel = LevelMinimal
		fmt.Println("⚠️ Degrading to LevelMinimal")
		s.mu.Unlock()
	}
	return result, err
}

func (s *DegradedService) handleMinimal(ctx context.Context) (string, error) {
	s.mu.Lock()
	s.requestsMinimal++
	s.mu.Unlock()

	// Return cached or fallback data
	// For demo, this would be cached response
	return "cached_response", nil
}

// GetStats returns current service statistics
func (s *DegradedService) GetStats() (full, degraded, minimal int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.requestsFull, s.requestsDegraded, s.requestsMinimal
}

// =============================================================================
// MAIN: Demonstration
// =============================================================================

func main() {
	fmt.Println("╔════════════════════════════════════════════════════════════════╗")
	fmt.Println("║  Chapter 5: Handling Infrastructure Variability Demo         ║")
	fmt.Println("║  (The Un-virtualized Ground)                                  ║")
	fmt.Println("╚════════════════════════════════════════════════════════════════╝")
	fmt.Println()

	// Create a client with high variance (simulating problematic infrastructure)
	// base: 100ms, variance: 500ms, failure rate: 10%
	client := NewInfrastructureClient("problematic-db", 100*time.Millisecond, 500*time.Millisecond, 0.1)

	// Demo 1: Naive approach - will hang
	fmt.Println("--- Demo 1: Naive Request (no timeout) ---")
	fmt.Println("⚠️  This would hang indefinitely in production")
	// result, _ := NaiveRequestWithoutTimeout(client) // Don't run this!

	// Demo 2: With timeout
	fmt.Println("\n--- Demo 2: Request with Timeout ---")
	result, err := RequestWithTimeout(client, 200*time.Millisecond)
	if err != nil {
		fmt.Printf("✓ Correctly handled: %v\n", err)
	} else {
		fmt.Printf("✓ Success: %s\n", result)
	}

	// Demo 3: Circuit Breaker
	fmt.Println("\n--- Demo 3: Circuit Breaker ---")
	cb := NewCircuitBreaker(3, 5*time.Second) // Open after 3 failures

	for i := 0; i < 10; i++ {
		err := cb.Execute(func() error {
			// 50% failure rate to trip circuit
			if rand.Float64() > 0.5 {
				return errors.New("simulated failure")
			}
			return nil
		})

		state := []string{"CLOSED", "OPEN", "HALF-OPEN"}[cb.GetState()]
		if err != nil {
			fmt.Printf("Request %d: FAILED - %v [Circuit: %s]\n", i+1, err, state)
		} else {
			fmt.Printf("Request %d: SUCCESS [Circuit: %s]\n", i+1, state)
		}

		if cb.GetState() == CircuitOpen {
			fmt.Println("→ Circuit opened! Further requests will fail fast.")
			break
		}
	}

	// Demo 4: Retry with Backoff
	fmt.Println("\n--- Demo 4: Retry with Exponential Backoff ---")
	client2 := NewInfrastructureClient("unreliable-service", 50*time.Millisecond, 100*time.Millisecond, 0.3)
	result, err = RetryWithBackoff(client2, DefaultRetryConfig())
	if err != nil {
		fmt.Printf("✓ All retries failed: %v\n", err)
	} else {
		fmt.Printf("✓ Success after retries: %s\n", result)
	}

	// Demo 5: Graceful Degradation
	fmt.Println("\n--- Demo 5: Graceful Degradation ---")
	client3 := NewInfrastructureClient("degrading-service", 100*time.Millisecond, 50*time.Millisecond, 0.4)
	service := NewDegradedService(client3)

	for i := 0; i < 10; i++ {
		result, err := service.HandleRequest(context.Background())
		if err != nil {
			fmt.Printf("Request %d: %v\n", i+1, err)
		} else {
			fmt.Printf("Request %d: %s\n", i+1, result)
		}
		time.Sleep(100 * time.Millisecond)
	}

	full, degraded, minimal := service.GetStats()
	fmt.Printf("\nStats - Full: %d, Degraded: %d, Minimal: %d\n", full, degraded, minimal)

	fmt.Println("\n✅ Demo complete!")
}
